File: driver/video_capture.py
#!/usr/bin/env python3
"""
video_capture.py - Simple video capture script using XDMA

This script captures video frames from the PCIe video capture card
and saves them as PNG images.
echo -ne '\x05\x00\x00\x00' | sudo dd of=/dev/xdma0_user bs=4 seek=1 count=1
sudo python3 video_capture.py -t -i -n 5 -v
Usage:
    sudo python3 video_capture.py [options]
"""

import os
import sys
import time
import argparse
import subprocess
import select
import numpy as np
from PIL import Image

# Configuration
FRAME_WIDTH = 1920
FRAME_HEIGHT = 1080
BYTES_PER_PIXEL = 4
FRAME_SIZE = FRAME_WIDTH * FRAME_HEIGHT * BYTES_PER_PIXEL

# Device paths
DMA_DEVICE = "/dev/xdma0_c2h_0"
USER_DEVICE = "/dev/xdma0_user"
EVENT_DEVICE_VSYNC_DEFAULT = "/dev/xdma0_events_1"  # often maps to usr_irq_req[1] (vsync falling)
EVENT_DEVICE_VSYNC_CANDIDATES = [
    EVENT_DEVICE_VSYNC_DEFAULT,
    "/dev/xdma0_events_0",
]

# Register offsets
REG_CONTROL = 0x04

# Control bits
CTRL_ENABLE = 0x01
CTRL_TEST_MODE = 0x04


def write_reg_dd(offset, value):
    """Write a 32-bit register using dd command"""
    # Convert value to bytes (little-endian)
    data = value.to_bytes(4, byteorder='little')
    
    # Use dd to write
    seek_blocks = offset // 4
    cmd = f"dd of={USER_DEVICE} bs=4 seek={seek_blocks} count=1 2>/dev/null"
    
    try:
        proc = subprocess.Popen(cmd, shell=True, stdin=subprocess.PIPE)
        proc.communicate(input=data)
        return proc.returncode == 0
    except Exception as e:
        print(f"Error writing register: {e}")
        return False


def enable_capture(test_mode=False):
    """Enable video capture"""
    ctrl = CTRL_ENABLE
    if test_mode:
        ctrl |= CTRL_TEST_MODE
    return write_reg_dd(REG_CONTROL, ctrl)


def disable_capture():
    """Disable video capture"""
    return write_reg_dd(REG_CONTROL, 0)


def wait_for_interrupt(fd_event): 
    """Wait for an interrupt event (blocking)"""
    try:
        data = os.read(fd_event, 4)
        if len(data) >= 4:
            return int.from_bytes(data, byteorder='little')
        return -1
    except OSError as e: 
        print(f"Error waiting for interrupt: {e}") 
        return -1 


def open_event_device(preferred_path=None):
    """Open an XDMA event device, returning (fd, path) or (None, None)."""
    candidates = []
    if preferred_path:
        candidates.append(preferred_path)
    candidates.extend(EVENT_DEVICE_VSYNC_CANDIDATES)

    tried = set()
    for path in candidates:
        if not path or path in tried:
            continue
        tried.add(path)
        try:
            fd = os.open(path, os.O_RDONLY)
            return fd, path
        except OSError:
            continue

    return None, None


def capture_frame_dd(output_file, size=FRAME_SIZE):
    """Capture one frame via DMA using dd command"""
    cmd = f"dd if={DMA_DEVICE} of={output_file} bs={size} count=1 2>/dev/null"
    try:
        result = subprocess.run(cmd, shell=True, capture_output=True)
        return result.returncode == 0
    except Exception as e:
        print(f"Error capturing frame: {e}")
        return False


def capture_frame(fd_dma, size=FRAME_SIZE, timeout_s=2.0, chunk_size=None): 
    """Capture one frame via DMA (read exactly `size` bytes). 
 
    Note: XDMA stream reads may return short reads even in blocking mode. 
    If we don't drain the full frame, the next capture will start mid-stream 
    and images will appear shifted/misaligned. 
    """ 
    if chunk_size is None:
        chunk_size = size  # default: try to read the whole frame per sys-call
    buf = bytearray() 
    read_calls = 0
    deadline = None if timeout_s is None else (time.monotonic() + float(timeout_s)) 
 
    try: 
        while len(buf) < size: 
            if deadline is not None:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    break
                rlist, _, _ = select.select([fd_dma], [], [], remaining)
                if not rlist:
                    continue

            to_read = min(size - len(buf), chunk_size) 
            chunk = os.read(fd_dma, to_read) 
            read_calls += 1
            if not chunk: 
                break 
            buf.extend(chunk) 
 
        return bytes(buf), read_calls 
    except OSError as e: 
        print(f"Error capturing frame: {e}") 
        return None, 0


def save_frame_png(filename, data, width=FRAME_WIDTH, height=FRAME_HEIGHT):
    """Save frame as PNG image (convert BGRX to RGB)"""
    if len(data) < width * height * 4:
        print(f"Warning: Incomplete frame data ({len(data)} bytes)")
        height = len(data) // (width * 4)
    
    if height == 0:
        print("Error: No valid frame data")
        return False
    
    frame = np.frombuffer(data[:height*width*4], dtype=np.uint8)
    frame = frame.reshape((height, width, 4))
    
    # Convert BGRX to RGB
    rgb = frame[:, :, 2::-1]  # Reverse first 3 channels
    
    img = Image.fromarray(rgb)
    img.save(filename)
    return True


def main():
    parser = argparse.ArgumentParser(description='PCIe Video Capture')
    parser.add_argument('-n', '--count', type=int, default=1,
                        help='Number of frames to capture (default: 1)')
    parser.add_argument('-o', '--output', default='frame',
                        help='Output file prefix (default: frame)')
    parser.add_argument('-t', '--test', action='store_true',
                        help='Use test pattern (color bar)')
    parser.add_argument('-i', '--interrupt', action='store_true', 
                        help='Use interrupt for frame sync') 
    parser.add_argument('--event', default=None,
                        help='XDMA event device path (default: auto-detect)')
    parser.add_argument('-d', '--delay', type=float, default=0.02, 
                        help='Delay between frames in seconds (default: 0.02)') 
    parser.add_argument('--chunk', type=int, default=FRAME_SIZE,
                        help='Max bytes per DMA read() call (default: frame size)') 
    parser.add_argument('--skip', type=int, default=0,
                        help='Discard N frames after enable (default: 0)') 
    parser.add_argument('-v', '--verbose', action='store_true', 
                        help='Verbose output') 
    args = parser.parse_args() 
    
    print("=== PCIe Video Capture ===")
    print(f"Resolution: {FRAME_WIDTH}x{FRAME_HEIGHT}")
    print(f"Frame size: {FRAME_SIZE} bytes")
    print(f"Mode: {'Test Pattern' if args.test else 'Video Input'}")
    print(f"Sync: {'Interrupt' if args.interrupt else 'Polling'}")
    print(f"Frames: {args.count}")
    print()
    
    fd_dma = None
    fd_event = None
    
    try:
        # Open DMA device
        fd_dma = os.open(DMA_DEVICE, os.O_RDONLY)
        print(f"Opened {DMA_DEVICE}")
        
        # Open event device if using interrupt 
        if args.interrupt: 
            fd_event, event_path = open_event_device(args.event)
            if fd_event is not None:
                print(f"Opened {event_path} for VSYNC interrupts")
            else:
                print("Warning: Could not open any XDMA event device, falling back to polling mode")
                args.interrupt = False
        
        # Enable capture
        print("Enabling video capture...")
        if not enable_capture(args.test): 
            print("Warning: Could not enable capture via register, trying anyway...") 
        time.sleep(0.1)  # Wait for video to stabilize 

        # Optional warm-up: discard a few frames to let the stream/driver settle.
        if args.skip > 0:
            if args.verbose:
                print(f"Discarding {args.skip} frame(s) for warm-up...")
            for s in range(args.skip):
                if args.interrupt and fd_event:
                    _ = wait_for_interrupt(fd_event)
                else:
                    time.sleep(args.delay)
                _data, _reads = capture_frame(fd_dma, chunk_size=args.chunk)
                if args.verbose:
                    got = 0 if _data is None else len(_data)
                    print(f"  Warm-up {s+1}/{args.skip}: {got} bytes, read() calls: {_reads}")
         
        # Capture frames 
        print(f"Capturing {args.count} frame(s)...") 
        start_time = time.time() 
        frames_saved = 0 
        
        for i in range(args.count):
            # Wait for frame sync
            if args.interrupt and fd_event:
                events = wait_for_interrupt(fd_event)
                if args.verbose:
                    print(f"  Interrupt received: {events}")
            else: 
                time.sleep(args.delay) 
             
            # Capture frame 
            data, read_calls = capture_frame(fd_dma, chunk_size=args.chunk) 
            if args.verbose:
                print(f"  DMA read() calls: {read_calls}")
             
            if data is None or len(data) == 0: 
                print(f"Warning: Failed to capture frame {i+1}") 
                continue 
            if len(data) != FRAME_SIZE:
                print(f"Error: Short/partial DMA read ({len(data)} bytes), aborting to avoid stream desync")
                break
            
            # Save frame
            filename = f"{args.output}_{i+1:04d}.png"
            if save_frame_png(filename, data):
                if args.verbose:
                    print(f"  Saved {filename} ({len(data)} bytes)")
                else:
                    print(f"\rCaptured {i+1}/{args.count} frames", end='')
                frames_saved += 1
        
        elapsed = time.time() - start_time
        print()
        print()
        print("=== Capture Complete ===")
        print(f"Frames captured: {frames_saved}")
        print(f"Elapsed time: {elapsed:.2f} seconds")
        if elapsed > 0:
            print(f"Average FPS: {frames_saved/elapsed:.2f}")
        
    except FileNotFoundError as e:
        print(f"Error: Device not found: {e}")
        print("Make sure the XDMA driver is loaded.")
        return 1
    except PermissionError as e:
        print(f"Error: Permission denied: {e}")
        print("Try running with sudo.")
        return 1
    except Exception as e:
        print(f"Error: {e}")
        import traceback
        traceback.print_exc()
        return 1
    finally:
        # Cleanup
        disable_capture()
        if fd_dma:
            os.close(fd_dma)
        if fd_event:
            os.close(fd_event)
    
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: driver/test/video_capture.c
/*
 * video_capture.c - Video capture application using XDMA with interrupt support
 *
 * This application captures video frames from the PCIe video capture card
 * using XDMA driver with interrupt-based frame synchronization.
 *
 * Interrupt mapping:
 *   - IRQ 0: VSYNC rising edge (frame start)
 *   - IRQ 1: VSYNC falling edge (active video start)
 *   - IRQ 2: Frame complete (DMA transfer done)
 *
 * Build: gcc -o video_capture video_capture.c -lpthread -O2
 * Usage: sudo ./video_capture [options]
 */

#include <errno.h>
#include <fcntl.h>
#include <pthread.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/mman.h>
#include <sys/time.h>
#include <unistd.h>


/* Configuration */
#define FRAME_WIDTH 1920
#define FRAME_HEIGHT 1080
#define BYTES_PER_PIXEL 4
#define FRAME_SIZE (FRAME_WIDTH * FRAME_HEIGHT * BYTES_PER_PIXEL)

#define NUM_BUFFERS 4 /* Number of frame buffers for ring buffer */
#define DMA_DEVICE "/dev/xdma0_c2h_0"
#define USER_DEVICE "/dev/xdma0_user"
#define EVENT_DEVICE_0 "/dev/xdma0_events_0" /* VSYNC rising */
#define EVENT_DEVICE_1 "/dev/xdma0_events_1" /* VSYNC falling */
#define EVENT_DEVICE_2 "/dev/xdma0_events_2" /* Frame complete */

/* Register offsets */
#define REG_VERSION 0x00
#define REG_CONTROL 0x04
#define REG_STATUS 0x08

/* Control bits */
#define CTRL_ENABLE (1 << 0)
#define CTRL_TEST_MODE (1 << 2)

/* Global variables */
static int running = 1;
static int fd_dma = -1;
static int fd_user = -1;
static int fd_event = -1;
static uint8_t *frame_buffers[NUM_BUFFERS];
static int current_buffer = 0;
static int frames_captured = 0;
static struct timeval start_time;

/* Signal handler */
void signal_handler(int sig) {
  printf("\nReceived signal %d, stopping...\n", sig);
  running = 0;
}

/* Write register */
int write_reg(int fd, uint32_t offset, uint32_t value) {
  if (lseek(fd, offset, SEEK_SET) < 0) {
    perror("lseek failed");
    return -1;
  }
  if (write(fd, &value, 4) != 4) {
    perror("write failed");
    return -1;
  }
  return 0;
}

/* Read register */
int read_reg(int fd, uint32_t offset, uint32_t *value) {
  if (lseek(fd, offset, SEEK_SET) < 0) {
    perror("lseek failed");
    return -1;
  }
  if (read(fd, value, 4) != 4) {
    perror("read failed");
    return -1;
  }
  return 0;
}

/* Enable video capture */
int enable_capture(int fd, int test_mode) {
  uint32_t ctrl = CTRL_ENABLE;
  if (test_mode) {
    ctrl |= CTRL_TEST_MODE;
  }
  return write_reg(fd, REG_CONTROL, ctrl);
}

/* Disable video capture */
int disable_capture(int fd) { return write_reg(fd, REG_CONTROL, 0); }

/* Wait for interrupt (blocking) */
int wait_for_interrupt(int fd_event, uint32_t *events) {
  ssize_t ret = read(fd_event, events, sizeof(*events));
  if (ret < 0) {
    if (errno == EINTR) {
      return 0; /* Interrupted by signal */
    }
    perror("read event failed");
    return -1;
  }
  return 1;
}

/* Capture one frame via DMA */
int capture_frame(int fd_dma, uint8_t *buffer, size_t size) {
  if (lseek(fd_dma, 0, SEEK_SET) < 0) {
    perror("lseek failed");
    return -1;
  }

  ssize_t total = 0;
  while (total < size) {
    ssize_t ret = read(fd_dma, buffer + total, size - total);
    if (ret < 0) {
      if (errno == EINTR)
        continue;
      perror("DMA read failed");
      return -1;
    }
    if (ret == 0)
      break; /* EOF or underflow */
    total += ret;
  }

  return total;
}

/* Save frame to file */
int save_frame(const char *filename, uint8_t *buffer, size_t size) {
  FILE *fp = fopen(filename, "wb");
  if (!fp) {
    perror("fopen failed");
    return -1;
  }

  size_t written = fwrite(buffer, 1, size, fp);
  fclose(fp);

  return (written == size) ? 0 : -1;
}

/* Save frame as PPM image */
int save_frame_ppm(const char *filename, uint8_t *buffer, int width,
                   int height) {
  FILE *fp = fopen(filename, "wb");
  if (!fp) {
    perror("fopen failed");
    return -1;
  }

  fprintf(fp, "P6\n%d %d\n255\n", width, height);

  /* Convert BGRX to RGB */
  for (int y = 0; y < height; y++) {
    for (int x = 0; x < width; x++) {
      int offset = (y * width + x) * 4;
      uint8_t b = buffer[offset + 0];
      uint8_t g = buffer[offset + 1];
      uint8_t r = buffer[offset + 2];
      fputc(r, fp);
      fputc(g, fp);
      fputc(b, fp);
    }
  }

  fclose(fp);
  return 0;
}

/* Print usage */
void print_usage(const char *prog) {
  printf("Usage: %s [options]\n", prog);
  printf("Options:\n");
  printf("  -n <count>    Number of frames to capture (0=continuous, "
         "default=10)\n");
  printf("  -o <prefix>   Output file prefix (default=frame)\n");
  printf("  -t            Use test pattern (color bar)\n");
  printf("  -i            Use interrupt for frame sync (default)\n");
  printf("  -p            Use polling for frame sync\n");
  printf("  -s            Save frames to files\n");
  printf("  -v            Verbose output\n");
  printf("  -h            Print this help\n");
}

int main(int argc, char *argv[]) {
  int opt;
  int frame_count = 10;
  int test_mode = 0;
  int use_interrupt = 1;
  int save_frames = 0;
  int verbose = 0;
  const char *output_prefix = "frame";

  /* Parse command line */
  while ((opt = getopt(argc, argv, "n:o:tipsvh")) != -1) {
    switch (opt) {
    case 'n':
      frame_count = atoi(optarg);
      break;
    case 'o':
      output_prefix = optarg;
      break;
    case 't':
      test_mode = 1;
      break;
    case 'i':
      use_interrupt = 1;
      break;
    case 'p':
      use_interrupt = 0;
      break;
    case 's':
      save_frames = 1;
      break;
    case 'v':
      verbose = 1;
      break;
    case 'h':
    default:
      print_usage(argv[0]);
      return (opt == 'h') ? 0 : 1;
    }
  }

  printf("=== PCIe Video Capture ===\n");
  printf("Resolution: %dx%d\n", FRAME_WIDTH, FRAME_HEIGHT);
  printf("Frame size: %d bytes\n", FRAME_SIZE);
  printf("Mode: %s\n", test_mode ? "Test Pattern" : "Video Input");
  printf("Sync: %s\n", use_interrupt ? "Interrupt" : "Polling");
  printf("Frames: %d%s\n", frame_count,
         frame_count == 0 ? " (continuous)" : "");
  printf("\n");

  /* Install signal handler */
  signal(SIGINT, signal_handler);
  signal(SIGTERM, signal_handler);

  /* Allocate frame buffers */
  for (int i = 0; i < NUM_BUFFERS; i++) {
    frame_buffers[i] = (uint8_t *)aligned_alloc(4096, FRAME_SIZE);
    if (!frame_buffers[i]) {
      perror("Failed to allocate frame buffer");
      goto cleanup;
    }
  }

  /* Open DMA device */
  fd_dma = open(DMA_DEVICE, O_RDONLY);
  if (fd_dma < 0) {
    perror("Failed to open DMA device");
    goto cleanup;
  }

  /* Open user device for register access */
  fd_user = open(USER_DEVICE, O_RDWR);
  if (fd_user < 0) {
    perror("Failed to open user device");
    goto cleanup;
  }

  /* Open event device for interrupts */
  if (use_interrupt) {
    fd_event = open(EVENT_DEVICE_2, O_RDONLY); /* Frame complete interrupt */
    if (fd_event < 0) {
      perror("Failed to open event device, falling back to polling");
      use_interrupt = 0;
    }
  }

  /* Read version */
  uint32_t version;
  if (read_reg(fd_user, REG_VERSION, &version) == 0) {
    printf("FPGA Version: 0x%08X\n", version);
  }

  /* Enable video capture */
  printf("Enabling video capture...\n");
  if (enable_capture(fd_user, test_mode) < 0) {
    fprintf(stderr, "Failed to enable capture\n");
    goto cleanup;
  }

  /* Wait for video to stabilize */
  usleep(100000); /* 100ms */

  /* Start timing */
  gettimeofday(&start_time, NULL);

  /* Main capture loop */
  printf("Starting capture...\n");
  while (running && (frame_count == 0 || frames_captured < frame_count)) {
    uint32_t events;
    int bytes;

    /* Wait for frame sync */
    if (use_interrupt) {
      int ret = wait_for_interrupt(fd_event, &events);
      if (ret < 0)
        break;
      if (ret == 0)
        continue; /* Interrupted */
    }

    /* Capture frame */
    uint8_t *buffer = frame_buffers[current_buffer];
    bytes = capture_frame(fd_dma, buffer, FRAME_SIZE);

    if (bytes < 0) {
      fprintf(stderr, "Capture failed\n");
      break;
    }

    if (bytes < FRAME_SIZE) {
      if (verbose) {
        printf("Warning: Partial frame %d bytes (expected %d)\n", bytes,
               FRAME_SIZE);
      }
    }

    frames_captured++;

    /* Save frame if requested */
    if (save_frames) {
      char filename[256];
      snprintf(filename, sizeof(filename), "%s_%04d.ppm", output_prefix,
               frames_captured);
      save_frame_ppm(filename, buffer, FRAME_WIDTH, FRAME_HEIGHT);
      if (verbose) {
        printf("Saved %s\n", filename);
      }
    }

    /* Progress */
    if (verbose || (frames_captured % 60 == 0)) {
      struct timeval now;
      gettimeofday(&now, NULL);
      double elapsed = (now.tv_sec - start_time.tv_sec) +
                       (now.tv_usec - start_time.tv_usec) / 1000000.0;
      double fps = frames_captured / elapsed;
      printf("\rCaptured %d frames (%.1f fps)   ", frames_captured, fps);
      fflush(stdout);
    }

    /* Next buffer */
    current_buffer = (current_buffer + 1) % NUM_BUFFERS;
  }

  /* Calculate final statistics */
  struct timeval end_time;
  gettimeofday(&end_time, NULL);
  double elapsed = (end_time.tv_sec - start_time.tv_sec) +
                   (end_time.tv_usec - start_time.tv_usec) / 1000000.0;

  printf("\n\n=== Capture Complete ===\n");
  printf("Frames captured: %d\n", frames_captured);
  printf("Elapsed time: %.2f seconds\n", elapsed);
  printf("Average FPS: %.2f\n", frames_captured / elapsed);
  printf("Data rate: %.2f MB/s\n",
         (frames_captured * FRAME_SIZE) / (elapsed * 1024 * 1024));

cleanup:
  /* Disable capture */
  if (fd_user >= 0) {
    disable_capture(fd_user);
    close(fd_user);
  }

  /* Close devices */
  if (fd_dma >= 0)
    close(fd_dma);
  if (fd_event >= 0)
    close(fd_event);

  /* Free buffers */
  for (int i = 0; i < NUM_BUFFERS; i++) {
    if (frame_buffers[i])
      free(frame_buffers[i]);
  }

  return 0;
}
